File: server/src/test/java/com/chatroom/server/ChatServerTest.java
package com.chatroom.server;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.io.*;
import java.net.Socket;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

class ChatServerTest {

    private ChatServer server;
    private ServerDisplay mockDisplay;
    private ExecutorService executor;
    private int serverPort;

    @BeforeEach
    void setUp() throws Exception {
        mockDisplay = message -> {};
        server = new ChatServer(0, mockDisplay);
        executor = Executors.newSingleThreadExecutor();
        
        executor.submit(() -> server.start());
        
        // Wait for server to start
        for (int i = 0; i < 20; i++) {
            serverPort = server.getPort();
            if (serverPort > 0) break;
            Thread.sleep(100);
        }
        assertTrue(serverPort > 0, "Server did not start");
    }

    @AfterEach
    void tearDown() {
        if (server != null) {
            server.stopServer();
        }
        if (executor != null) {
            executor.shutdownNow();
            try {
                if (!executor.awaitTermination(2, TimeUnit.SECONDS)) {
                    executor.shutdownNow();
                }
            } catch (InterruptedException e) {
                executor.shutdownNow();
            }
        }
    }

    @Test
    @Timeout(5)
    void testServerStartsOnPort() {
        assertTrue(server.getPort() > 0, "Server should be running on a valid port");
    }

    @Test
    @Timeout(10)
    void testClientConnection() throws Exception {
        CountDownLatch latch = new CountDownLatch(1);
        
        try (Socket socket = new Socket("localhost", serverPort);
             ObjectOutputStream out = new ObjectOutputStream(socket.getOutputStream());
             ObjectInputStream in = new ObjectInputStream(socket.getInputStream())) {
            
            out.writeObject("TestUser");
            out.flush();
            
            Thread reader = new Thread(() -> {
                try {
                    Object received = in.readObject();
                    if (received != null) {
                        latch.countDown();
                    }
                } catch (Exception e) {
                    // ignore
                }
            });
            reader.start();
            reader.join(3000);
            
            assertEquals(0, latch.getCount(), "Should receive response from server");
        }
    }
}

File: server/src/test/kotlin/com/chatroom/server/ChatServerTest.kt
package com.chatroom.server

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import java.net.Socket
import java.io.ObjectOutputStream
import java.io.ObjectInputStream
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

class ChatServerTest {

    @Test
    fun testServerBroadcast() {
        val display = object : ServerDisplay {
            override fun showMessage(message: String) {}
        }
        val server = ChatServer(0, display)

        val serverThread = thread {
            server.start()
        }

        // Wait for server to be ready
        var port = -1
        for (i in 0..20) {
            port = server.getPort()
            if (port != -1 && port != 0) break
            Thread.sleep(100)
        }
        assertNotEquals(-1, port)
        assertNotEquals(0, port)

        val latch = CountDownLatch(1)

        // Client 1
        thread {
            try {
                val socket = Socket("localhost", port)
                val out = ObjectOutputStream(socket.getOutputStream())
                out.flush()
                val input = ObjectInputStream(socket.getInputStream())

                // Send username
                out.writeObject("User1")
                out.flush()

                // Wait for broadcast
                while (true) {
                    val msg = input.readObject().toString()
                    if (msg.contains("Hello World")) {
                        latch.countDown()
                        break
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        // Client 2 (Sender)
        thread {
             Thread.sleep(1000) // wait for Client 1 to connect
             try {
                val socket = Socket("localhost", port)
                val out = ObjectOutputStream(socket.getOutputStream())
                out.flush()
                val input = ObjectInputStream(socket.getInputStream())

                out.writeObject("User2")
                out.flush()

                out.writeObject("@EE@|Hello World")
                out.flush()
             } catch (e: Exception) {
                e.printStackTrace()
             }
        }

        assertTrue(latch.await(10, TimeUnit.SECONDS))
    }
}
